src/effects/throw_spiny.c

```diff
@@ -1,14 +1,171 @@
+
+    if (time <= 0) {
+        spinyObject->life = 1000;
+    } else {
+        spinyObject->life = time + 60;
+    }
+
+    temp_f8 = arg5 - arg2;
+    temp_f4 = time;
+    gravity = (temp_f8 / temp_f4) - (time * -0.10000000149011612); //TODO: fix this weird number
+    spinyObject->timeUntilFall = time;
+    spinyObject->rgba = 255;
+    spinyObject->unk_10 = arg1;
+    spinyObject->unk_14 = arg2;
+    spinyObject->unk_18 = arg3;
+    spinyObject->unk_1C = arg4;
+    spinyObject->unk_20 = arg5;
+    spinyObject->unk_24 = arg6;
+    spinyObject->unk_44 = (arg4 - arg1) / temp_f4;
+    spinyObject->unk_4C = (arg6 - arg3) / temp_f4;
+    spinyObject->pos.x = arg1;
+    spinyObject->pos.y = arg2;
+    spinyObject->pos.z = arg3;
+    spinyObject->gravity = gravity;
+
+    if (temp_f8 < 0.0f) {
+        spinyObject->unk_64 = 1;
+    } else {
+        spinyObject->unk_64 = 0;
+    }
+
+    spinyObject->unk_40 = arg7;
+    spinyObject->unk_30 = 70;
+    spinyObject->unk_34 = 180;
+    spinyObject->unk_38 = 120;
+    spinyObject->yaw = shim_rand_int(360);
+    spinyObject->rotationSpeed = shim_rand_int(10) + 5;
+    spinyObject->state = -1;
+    spinyObject->xScale = 1.0f;
+    spinyObject->yScale = 1.0f;
+    return effect;
+}
 
 void throw_spiny_init(void) {
 }
 
-INCLUDE_ASM(s32, "effects/throw_spiny", throw_spiny_update);
+void throw_spiny_update(EffectInstance* effectInstance) {
+    ThrowSpinyFXData* spinyObject = effectInstance->data;
+    u32 state;
+    f32 gravity;
+    s32 tempVar;
+
+    if (effectInstance->flags & EFFECT_INSTANCE_FLAGS_10) {
+        effectInstance->flags &= ~EFFECT_INSTANCE_FLAGS_10;
+        spinyObject->life = 16;
+    }
+
+    if (spinyObject->life < 1000) {
+        spinyObject->life--;
+    }
+
+    spinyObject->lifeDuration++;
+
+    if (spinyObject->life < 0) {
+        shim_remove_effect((EffectInstance*)effectInstance);
+        return;
+    }
+
+    tempVar = spinyObject->lifeDuration;
+    state = spinyObject->state;
+
+    if (state < 7) {
+        spinyObject->xScale = D_E00C8718[state] * 0.01f;
+        spinyObject->yScale = D_E00C8720[state] * 0.01f;
+        spinyObject->state += 1;
+        spinyObject->yaw -= 40.0f;
+    } else {
+        spinyObject->gravity += -0.2f;
+        spinyObject->pos.x += spinyObject->unk_44;
+        spinyObject->pos.y += spinyObject->gravity;
+        spinyObject->pos.z += spinyObject->unk_4C;
+        spinyObject->yaw += spinyObject->rotationSpeed;
+    }
+
+    if ((tempVar - 1) == spinyObject->timeUntilFall) {
+        spinyObject->state = 0;
+        spinyObject->gravity = -spinyObject->gravity;
+        spinyObject->unk_44 = spinyObject->unk_44;
+        spinyObject->rotationSpeed = -4.0f;
+        return;
+    }
+
+    gravity = spinyObject->gravity;
+
+    if ((gravity < 0.0f) && (spinyObject->pos.y < 100.0 / 7.0)) {
+        spinyObject->pos.y = 100.0f / 7.0f;
+        spinyObject->rotationSpeed = -20.0f;
+        spinyObject->gravity = gravity - gravity;
```

Comment:
wot... this is 0 right? but I assume 0 doesn't work lol

Reply:
for some reason assigning spinyObject->gravity to 0.0f doesn't match